File: HW4_1/src/main/java/com/homework/hw4_1/MainActivity.java
package com.homework.hw4_1;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.SearchView;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity {
    public static final String NEW_CONTACT = "NewContact";
    public static final String CONTACT_FOR_CHANGES = "ContactForChanges";
    public static final String CONTACT_NUMBER = "ContactNumber";
    public static final String MODIFIED_CONTACT = "ModifiedContact";
    private List<Contact> contactList = new ArrayList<>();  //
    private MenuItem menuItem;
    private SearchView searchView;
    private ContactListAdapter adapter;
    private RecyclerView recyclerView;

//    @Override
//    public boolean onCreateOptionsMenu(Menu menu) {
//        getMenuInflater().inflate(R.menu.menu, menu);
//        menuItem = menu.findItem(R.id.search_on_menu);
//        searchView = (SearchView) menuItem.getActionView();

//        searchView.findViewById(R.id.searchContact);
//        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
//            @Override
//            public boolean onQueryTextSubmit(String query) {
//                return false;
//            }
//
//            @Override
//            public boolean onQueryTextChange(String newText) {
//                adapter.getFilter().filter(newText);
//                return false;
//            }
//        });
//        return true;
//    }

    public interface ListContactActionListener {
        void onContactClicked(int number);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        recyclerView = findViewById(R.id.recyclerView);

        recyclerView.setAdapter(new ContactListAdapter(contactList, new ListContactActionListener() {
            @Override
            public void onContactClicked(int number) {
                Intent intent = new Intent(
                        MainActivity.this, ActivityEditContact.class);
                intent.putExtra(CONTACT_FOR_CHANGES, contactList.get(number));
                intent.putExtra(CONTACT_NUMBER, number);
                startActivityForResult(intent, 456);

            }
        }));
        recyclerView.setLayoutManager(
                new LinearLayoutManager(this, RecyclerView.VERTICAL, false));

        findViewById(R.id.buttonAddContact).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivityForResult(
                        new Intent(MainActivity.this, ActivityAddContact.class),
                        123);
            }
        });

        searchView.findViewById(R.id.searchContact);
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                return false;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                adapter.getFilter().filter(newText);
                return false;
            }
        });

    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        if (requestCode == 123 && resultCode == Activity.RESULT_OK && data != null) {
            contactList.add((Contact) data.getSerializableExtra(NEW_CONTACT));
            adapter.notifyItemChanged(contactList.size() - 1);
        } else if (requestCode == 456 && resultCode == Activity.RESULT_OK && data != null) {
            contactList.set(data.getIntExtra(CONTACT_NUMBER, 0),
                    (Contact) data.getSerializableExtra(MODIFIED_CONTACT));
            adapter.notifyDataSetChanged();
        }
        super.onActivityResult(requestCode, resultCode, data);
    }
}

// как убрать ошибку "Dependent features configured but no package ID was set." ?

File: hw5_2/src/main/java/com/homework/hw5_2/MainActivity.kt
package com.homework.hw5_2

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.SearchView
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView

open class MainActivity : AppCompatActivity() {
    protected val dBManager = com.homework.hw5_2.db.DBManager(this)
    private lateinit var searchView: SearchView
    private lateinit var adapter: ContactListAdapter
    private lateinit var recyclerView: RecyclerView

    interface ListContactActionListener {
        fun onContactClicked(number: Int)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        recyclerView = findViewById(R.id.recyclerView)
        dBManager.openDB()
        adapter = ContactListAdapter(
                object : ListContactActionListener {
                    override fun onContactClicked(number: Int) {
                        val intent = Intent(
                                this@MainActivity, ActivityEditContact::class.java)
                        intent.putExtra(CONTACT_NUMBER, number + 1)
                        startActivityForResult(intent, 456)
                    }
                })
        recyclerView.adapter = adapter
        recyclerView.layoutManager = LinearLayoutManager(this, RecyclerView.VERTICAL, false)

        findViewById<View>(R.id.buttonAddContact).setOnClickListener {
            startActivityForResult(
                    Intent(this@MainActivity, ActivityAddContact::class.java),
                    123)
        }
        adapter.addItems(dBManager.readDBData())
        searchView = findViewById(R.id.searchContact)
        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {

            override fun onQueryTextSubmit(query: String): Boolean {
                return false
            }

            override fun onQueryTextChange(newText: String): Boolean {
                adapter.filter.filter(newText)
                return false
            }
        })
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        dBManager.openDB()
        if (requestCode == 123 && resultCode == RESULT_OK && data != null) {
            adapter.addItems(dBManager.readDBData())
        } else if (requestCode == 456 && resultCode == RESULT_OK && data != null) {
            adapter.addItems(dBManager.readDBData())
        }
        super.onActivityResult(requestCode, resultCode, data)
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        dBManager.openDB()
        adapter.addItems(dBManager.readDBData())
        super.onRestoreInstanceState(savedInstanceState)
    }

    override fun onDestroy() {
        super.onDestroy()
        dBManager.closeDB()
    }

    companion object {
        const val NEW_CONTACT = "NewContact"
        const val CONTACT_NUMBER = "ContactNumber"
        const val MODIFIED_CONTACT = "ModifiedContact"
    }
}
